test/e2e/network/service.go

```diff
@@ -293,36 +294,48 @@ var _ = SIGDescribe("Services", func() {
 
 		ginkgo.By("Picking multiple nodes")
 		nodes := framework.GetReadySchedulableNodesOrDie(f.ClientSet)
-
-		if len(nodes.Items) == 1 {
-			framework.Skipf("The test requires two Ready nodes on %s, but found just one.", framework.TestContext.Provider)
+		nodeCounts := len(nodes.Items)
+		if nodeCounts < 2 {
+			framework.Skipf("The test requires atlest two ready nodes on %s, but found %v", framework.TestContext.Provider, nodeCounts)
 		}
 
-		node1 := nodes.Items[0]
-		node2 := nodes.Items[1]
-
 		ginkgo.By("Creating a webserver pod be part of the TCP service which echoes back source ip")
 		serverPodName := "echoserver-sourceip"
-		jig.LaunchEchoserverPodOnNode(f, node1.Name, serverPodName)
+		jig.LaunchEchoserverPod(f, serverPodName)
 		defer func() {
 			e2elog.Logf("Cleaning up the echo server pod")
 			err := cs.CoreV1().Pods(ns).Delete(serverPodName, nil)
-			framework.ExpectNoError(err, "failed to delete pod: %s on node: %s", serverPodName, node1.Name)
+			framework.ExpectNoError(err, "failed to delete pod: %s on node", serverPodName)
 		}()
 
 		// Waiting for service to expose endpoint.
 		err := e2eendpoints.ValidateEndpointsPorts(cs, ns, serviceName, e2eendpoints.PortsByPodName{serverPodName: {servicePort}})
 		framework.ExpectNoError(err, "failed to validate endpoints for service %s in namespace: %s", serviceName, ns)
 
-		ginkgo.By("Retrieve sourceip from a pod on the same node")
-		sourceIP1, execPodIP1 := execSourceipTest(f, cs, ns, node1.Name, serviceIP, servicePort)
-		ginkgo.By("Verifying the preserved source ip")
-		framework.ExpectEqual(sourceIP1, execPodIP1)
+		ginkgo.By("Creating pause pod deployment")
+		deployment := jig.CreatePausePodDeployment("pause-pod", ns, int32(nodeCounts))
```

Comment:
I don't think we should do this from *every* node. What if someone wants to run the conformance suite against a cluster with 3000 nodes? This would be a very long and expensive test. Similarly, we should restrict the NodePort checks to just a couple nodes. Can we go back to using just 2 nodes, and doing all the checks with just two nodes? Then it's easy to use nodeSelector to ensure scheduling on different nodes, and we don't run into issues with massive clusters. Plus, there can be tainted nodes, etc. I guess that's an issue for using nodeSelectors too - if we pick one and its tainted it won't schedule.

We could use pod anti-affinity and then verify they are scheduled on two different pods, and error out if not. That gets around the taints causing a problem.

So...let's go with that - use anti-affinity but only 2 pods.

Reply:
Thanks for pointing this gotcha. Using podAntiAffinity with HARD requirement to schedule pause pod in different node and considering only 2 nodes.